modules/programs/afew.nix

```diff
@@ -6,21 +6,134 @@ let
+  reorderFilters = filters:
+    (dag.topoSort (mkDag filters)).result;
+
+  /* Convert a topologically sorted list of DAG nodes to a list of INI
+     sections.  The section name is appended with an index number to
+     prevent clashes, as is required by afew.
+  */
+  toOrderedINI = {
+    mkKeyValue ? mkKeyValueDefault {} "="
+  }: filters:
+    let
+      mkSection = iota: node: ''
+        [${node.data.name}.${toString iota}]
+        ${toKV node.data.value}
+      '';
+      toKV = attrs: generators.toKeyValue { inherit mkKeyValue; } attrs;
+    in
+      concatStringsSep "\n" (imap0 mkSection filters);
+
 in
 
 {
   options.programs.afew = {
     enable = mkEnableOption "the afew initial tagging script for Notmuch";
 
+    defaultBefore = mkOption {
+      type = types.listOf types.string;
+      default = ["@end"];
+      description = ''
+        List of node names in the DAG that filters should appear
+        before by default.
+      '';
+    };
+
+    defaultAfter = mkOption {
+      type = types.listOf types.string;
+      default = ["@begin"];
+      description = ''
+        List of node names in the DAG that filters should appear after
+        by default.
+      '';
+    };
+
+    defaultFilters = mkOption {
+      type = types.attrsOf types.attrs;
+      default = {
+        "@begin" = { message = "BEGIN"; after = []; before = []; };
+        "@end" = { message = "END"; after = ["@begin"]; before = []; };
+      };
+      description = ''
+        An initial value for the filters DAG. It should only contain
+        "dummy" nodes. Those dummy nodes are conventionally prefixed
+        with @.
+      '';
+    };
+
+    filters = mkOption {
+      type = types.attrsOf types.attrs;
+      default = {
+        spam = { class = "SpamFilter"; before = ["@begin"]; after = []; };
+        killThreads = { class = "KillThreadsFilter"; before = ["@begin"]; after = []; };
+        listMails = { class = "ListMailsFilter"; before = ["@begin"]; after = []; };
+        archiveSentMails = { class = "ArchiveSentMailsFilter"; before = ["@begin"]; after = []; };
+        inbox = { class = "InboxFilter"; after = ["@end"]; before = []; };
+      };
+      description = ''
+        Filters added to afew configuration file. Available configuration
+        options are described in the afew manual: <link
+        xlink:href="https://afew.readthedocs.io/en/latest/configuration.html" />.
+      '';
+      example = literalExample ''
+        {
+          spam = {
+            class = "SpamFilter";
+            before = ["@begin"];
+            after = [];
+          };
+
+          boss = {
+            query = "from:pointyheaded@boss.com";
+            tags = ["-new" "+boss"];
+            message = "Message from above";
+          };
+
+          inbox = {
+            class = "InboxFilter";
+            before = [];
+            after = ["@end"];
+          };
+        }
+      '';
+    };
+
     extraConfig = mkOption {
       type = types.lines;
```

Comment:
I'd suggest to change this to `types.nullOr types.lines` and the default to `null`. Then add
```
warnings = optional (cfg.extraConfig != null)
  "'programs.afew.extraConfig' is deprecated, use 'programs.afew.filters' instead.";
```
and change
```
      ${cfg.extraConfig}
      ${toOrderedINI { mkKeyValue = mkIniKeyValue; } (reorderFilters (cfg.defaultFilters // cfg.filters))}
```
to
```
      ${
        if (cfg.extraConfig != null)
        then cfg.extraConfig
        else
          toOrderedINI { mkKeyValue = mkIniKeyValue; }
          (reorderFilters
          (defaultFilters // cfg.filters))
      }
```
where `defaultFilters` should be `cfg.defaultFilters` if it is not possible to hard code them.

Perhaps this will allow sufficient backwards compatibility and won't cause issues of odd mixing of configurations.

Reply:
This looks great, however I'm not familiar enough with the nix expression language or home-manager to know how to use `warnings`. Can you point me to the documentation or an example?